File: app/src/test/kotlin/com/saatxi/eatapp/data/local/LinkValidationTest.kt
package com.saatxi.eatapp.data.local

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * These two functions stand between a user-entered value and an
 * `Intent.ACTION_VIEW`, so the rejection cases matter more than the happy
 * path: a scheme that slipped through here would let a restaurant's own data
 * choose what the app launches.
 *
 * Plain Kotlin, no Robolectric — the rules deliberately don't touch `Uri`.
 */
class LinkValidationTest {

    @Test
    fun `keeps an http or https url as written`() {
        assertEquals("https://example.com", normalizeWebsite("https://example.com"))
        assertEquals("http://example.com/menu", normalizeWebsite("http://example.com/menu"))
    }

    @Test
    fun `assumes https for a bare host, which is how people tend to type it`() {
        assertEquals("https://example.com", normalizeWebsite("example.com"))
        assertEquals("https://example.com/menu", normalizeWebsite("example.com/menu"))
    }

    @Test
    fun `matches the scheme case-insensitively`() {
        assertEquals("HTTPS://example.com", normalizeWebsite("HTTPS://example.com"))
    }

    @Test
    fun `trims surrounding whitespace`() {
        assertEquals("https://example.com", normalizeWebsite("  https://example.com  "))
    }

    @Test
    fun `treats a missing, empty or blank value as no website`() {
        assertNull(normalizeWebsite(null))
        assertNull(normalizeWebsite(""))
        assertNull(normalizeWebsite("   "))
    }

    /** The reason this file exists. */
    @Test
    fun `rejects every scheme that is not http or https`() {
        assertNull(normalizeWebsite("javascript:alert(1)"))
        assertNull(normalizeWebsite("intent://scan/#Intent;scheme=zxing;end"))
        assertNull(normalizeWebsite("file:///data/data/com.saatxi.eatapp/databases/eatapp.db"))
        assertNull(normalizeWebsite("content://com.android.contacts/contacts"))
        assertNull(normalizeWebsite("data:text/html,<script>alert(1)</script>"))
        assertNull(normalizeWebsite("eatapp://whatever"))
    }

    @Test
    fun `accepts an instagram handle with or without the at sign`() {
        assertEquals("cal_ferran", normalizeInstagramHandle("cal_ferran"))
        assertEquals("cal_ferran", normalizeInstagramHandle("@cal_ferran"))
        assertEquals("cal.ferran99", normalizeInstagramHandle("  @cal.ferran99  "))
    }

    @Test
    fun `rejects a handle with characters instagram does not allow`() {
        assertNull(normalizeInstagramHandle("cal ferran"))
        assertNull(normalizeInstagramHandle("cal/ferran"))
        assertNull(normalizeInstagramHandle("cal-ferran"))
        // A full URL is not a handle: storing one would hand user input control
        // over the scheme, which is exactly what this design avoids.
        assertNull(normalizeInstagramHandle("https://instagram.com/cal_ferran"))
    }

    @Test
    fun `rejects an empty handle or one longer than instagram allows`() {
        assertNull(normalizeInstagramHandle(null))
        assertNull(normalizeInstagramHandle(""))
        assertNull(normalizeInstagramHandle("@"))
        assertEquals("a".repeat(30), normalizeInstagramHandle("a".repeat(30)))
        assertNull(normalizeInstagramHandle("a".repeat(31)))
    }

    @Test
    fun `builds the profile url from the handle rather than parsing one`() {
        assertEquals("https://instagram.com/cal_ferran", instagramUrl("cal_ferran"))
    }
}

File: test/core/utils/link_validation_test.dart
import 'package:eatapp/core/utils/link_validation.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('normalizeWebsite', () {
    test('assumes https for a bare host', () {
      expect(normalizeWebsite('example.com'), 'https://example.com');
      expect(normalizeWebsite('www.example.com/path?a=1'), 'https://www.example.com/path?a=1');
    });

    test('keeps an http or https URL as it was', () {
      expect(normalizeWebsite('https://example.com'), 'https://example.com');
      expect(normalizeWebsite('http://example.com'), 'http://example.com');
      expect(
        normalizeWebsite('HTTPS://EXAMPLE.COM'),
        'HTTPS://EXAMPLE.COM',
        reason: 'only the scheme is lowercased for the check, not the value',
      );
    });

    test('trims, and treats blank as absent', () {
      expect(normalizeWebsite('  https://example.com  '), 'https://example.com');
      expect(normalizeWebsite(null), isNull);
      expect(normalizeWebsite(''), isNull);
      expect(normalizeWebsite('   '), isNull);
    });

    test('refuses every scheme it does not recognise', () {
      // The whole point of the whitelist: these would otherwise be handed to
      // the system as an actionable URI.
      expect(normalizeWebsite('javascript:alert(1)'), isNull);
      expect(normalizeWebsite('intent://scan/#Intent;scheme=zxing;end'), isNull);
      expect(normalizeWebsite('file:///etc/passwd'), isNull);
      expect(normalizeWebsite('content://com.other.app/data'), isNull);
      expect(normalizeWebsite('mailto:someone@example.com'), isNull);
      expect(normalizeWebsite('ftp://example.com'), isNull);
      expect(normalizeWebsite('data:text/html,<script>'), isNull);
    });

    test('refuses a bare host carrying a port, as the Android app does', () {
      // Carried over deliberately rather than quietly improved: the scheme
      // pattern sees "example.com:" and reads it as an unknown scheme, so
      // "example.com:8080" is rejected. Behavioural parity matters more here
      // than fixing the wart.
      expect(normalizeWebsite('example.com:8080'), isNull);
      expect(normalizeWebsite('http://example.com:8080'), 'http://example.com:8080');
    });
  });

  group('normalizeInstagramHandle', () {
    test('accepts a handle with or without the leading @, and trims', () {
      expect(normalizeInstagramHandle('saatxi'), 'saatxi');
      expect(normalizeInstagramHandle('@saatxi'), 'saatxi');
      expect(normalizeInstagramHandle('  @saatxi  '), 'saatxi');
      expect(normalizeInstagramHandle('saatxi.dev'), 'saatxi.dev');
      expect(normalizeInstagramHandle('user_name'), 'user_name');
    });

    test('is exactly as long as Instagram allows, and not a character longer', () {
      final String thirty = 'a' * 30;
      final String thirtyOne = 'a' * 31;

      expect(normalizeInstagramHandle(thirty), thirty);
      expect(normalizeInstagramHandle(thirtyOne), isNull);
    });

    test('refuses a URL, a stray @ and any character outside the handle set', () {
      expect(normalizeInstagramHandle('https://instagram.com/saatxi'), isNull);
      expect(normalizeInstagramHandle('instagram.com/saatxi'), isNull);
      expect(normalizeInstagramHandle('@@saatxi'), isNull);
      expect(normalizeInstagramHandle('with-dash'), isNull);
      expect(normalizeInstagramHandle('with space'), isNull);
    });

    test('treats blank as absent', () {
      expect(normalizeInstagramHandle(null), isNull);
      expect(normalizeInstagramHandle(''), isNull);
      expect(normalizeInstagramHandle('@'), isNull);
      expect(normalizeInstagramHandle('   '), isNull);
    });

    test('the built URL is the base plus a handle that was already validated', () {
      expect(instagramUrl('saatxi'), 'https://instagram.com/saatxi');
    });
  });
}
